realm/rfc/0001-backlink-query-syntax.md

```diff
@@ -0,0 +1,249 @@
+classes, it should be possible to query backlinks without them being defined there.
+
+# Motivation
+
+Backlinks are special in Realm. They are not defined explicitly in the schema,
+but are automatically created when defining Links and LinkLists (forward links).
+
+This creates some conflicts with the current query system as it rely on all 
+fields being defined by the schema. This is called *named backlinks*.
+
+This proposal introduces a new syntax for queries so it is possible to query
+across a backlink even though it hasn't been defined by a model class. This is
+called *unnamed backlinks*.
+
+Realm Objective C/Swift shipped with only support for named backlinks. This was mostly
+prompted by restrictions in NSPredicate, which do not have a syntax for unnamed backlinks.
+
+Realm Java does not have this constraint and at the same time the Java Migration/Dynamic API
+is more full-fledged and expected to have the same feature set as the static API due to how the
+query api works.
+
+Adding support for the unnamed variant now will also most likely save us time documenting the
+difference as well as reduce the amount of support we have to do.
+
+Any design should ideally solve all these constraints:
+
+- Should support a list of backlink objects.
+
+- Should support a single backlink object, enforcing the constraint that only 1 object
+  can point to it. This is not supported by Core, but it is a natural extension.
+
+- Should be usable from the Dynamic API.
+
+- Should either support polymorphic objects or not prevent them from being 
+  implemented.
+
+- Ideally it should work for both the pure text-based variant of the query language in the
+  Object Store as well as the current mixed style found in Realm Java.
+  
+
+# Detailed design
+
+**Example model classes:**
+
+```
+public class Person extends RealmObject {
+	private String name
+	private Dog favoriteDog;
+	private RealmList<Dog> dogs;
+}
+
+public class Dog extends RealmObject {
+	// Named backlinks
+	@LinkingObjects("favoriteDog")
+	private RealmResults<Person> favoriteOwner;
+}
+```
+
+**Introduce the following query extension**
+
+- Use `linkingObjects(x.z).y` to search field Y on type X where X is a backlink object through
+  field Z. This is a supplement to defining the named backlinks, which will continue to work as normal
+  link queries: `favoriteOwner.name`
+- No changes to the migration API.
+- No changes to the schema definition.
+
+Example:
+
+```
+// Named backlink queries
+RealmResults<Dog> dogs = realm.where(Dog.class).equalTo("favoriteOwner.name", "John").findAll();
+
+// Unnamed backlink queries
+RealmResults<Dog> dogs = realm.where(Dog.class).equalTo("linkingObjects(Person.dogs).name", "John").findAll();
+RealmResults<DynamicRealmObjcet> dogs = dynamicRealm.where("Dog").equalTo("linkingObjects(Person.dogs).name", "John").findAll();
+
+// Example in a purely text based variant
+"ANY linkingObjects(Person.dogs).name = 'John'"
```

Comment:
The cost is that we have to parse the query (the string in `equalTo()` in the snippet above). If we decide to go down that route, I think it is better to use the query language implemented in Object Store.